File: frontend/src/types/request.ts
export type RequestStatus =
  | "NEW"
  | "IN_PROGRESS"
  | "CLARIFICATION_NEEDED"
  | "RESOLVED"
  | "CLOSED";

export type RequestPriority = "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";
export type ActorRole = "citizen" | "worker";

export interface CitizenRequest {
  id: number;
  title: string;
  description: string;
  category_id: number;
  priority: RequestPriority;
  status: RequestStatus;
  citizen_first_name: string;
  citizen_last_name: string;
  assigned_to_user_id?: number | null;
  assigned_to_display_name?: string | null;
  resolved_at?: string | null;
  closed_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface RequestStatusHistoryItem {
  id: number;
  request_id: number;
  from_status: RequestStatus | null;
  to_status: RequestStatus;
  changed_by_user_id: number;
  changed_by_display_name?: string | null;
  change_note?: string | null;
  changed_at: string;
}

export interface RequestCommentItem {
  id: number;
  request_id: number;
  author_user_id?: number | null;
  author_role: string;
  author_display_name: string;
  comment_text: string;
  created_at: string;
}

export interface RequestDetailResponse {
  request: CitizenRequest;
  comments: RequestCommentItem[];
  status_history: RequestStatusHistoryItem[];
}

export interface CreateRequestPayload {
  title: string;
  description: string;
  category_id: number;
  priority: RequestPriority;
  citizen_first_name: string;
  citizen_last_name: string;
}

File: backend/app/schemas/request.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.schemas.comment import CommentRead
from app.schemas.enums import RequestPriority, RequestStatus


class RequestCreate(BaseModel):
    creator_staff_profile_id: int | None = None
    title: str
    description: str
    category_id: int
    priority: RequestPriority
    citizen_first_name: str
    citizen_last_name: str


class RequestStatusUpdate(BaseModel):
    actor_staff_profile_id: int
    to_status: RequestStatus
    change_note: str | None = None


class RequestClaim(BaseModel):
    actor_staff_profile_id: int


class RequestRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str
    category_id: int
    priority: RequestPriority
    status: RequestStatus
    citizen_first_name: str
    citizen_last_name: str
    assigned_to_staff_profile_id: int | None
    assigned_to_display_name: str | None = None
    resolved_at: datetime | None
    closed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class RequestStatusHistoryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    request_id: int
    from_status: RequestStatus | None
    to_status: RequestStatus
    changed_by_staff_profile_id: int
    changed_by_display_name: str | None = None
    change_note: str | None
    changed_at: datetime


class RequestDetail(BaseModel):
    request: RequestRead
    comments: list[CommentRead]
    status_history: list[RequestStatusHistoryRead]
